web/src/Tree.js

```diff
@@ -0,0 +1,118 @@
+import React, { Component } from 'react';
+
+function Field(props) {
+    if (!props.value) {
+        return null;
+    } else if (props.length > 40) {
+        return [<p className="font-weight-bold">{props.name}</p>, <p>{props.value}</p>];
+    } else {
+        return (<p><span className="font-weight-bold">{props.name}: </span>{props.value}</p>);
+    }
+}
+
+function Email(props) {
+    const email = props.email;
+    return (<li className="list-group-item">
+        <Field name="Email" value={email.name}/>
+        <Field name="Subject" value={email.subject}/>
+        <Field name="Body" value={email.body}/>
+        <Field name="Attached Documents" value={email.documents.map(d => d.name).join(', ')}/>
+        <Field name="Generated Templates" value={email.templates.map(t => t.name + " of type " + t.document_type.name).join(', ')}/>
+        <Field name="Recipients" value={email.recipients.map(r => r.name).join(', ')}/>
+        <Field name="Required Partners" value={email.required.map(r => r.name).join(', ')}/>
+        <Field name="Excluded Partners" value={email.excluded.map(r => r.name).join(', ')}/>
+        </li>);
+}
+
+function Affect(props) {
+        return (<span>
+            <span className="font-weight-bold">On</span> {props.reason}
+            <span className="font-weight-bold"> {props.affect.type}</span> {props.affect.action.name}
+        </span>);
+}
+
+function Action(props) {
+    const action = props.action;
+    function affectLi(reason) {
+        return (a => (<li className="list-group-item">
+            <Affect key={a.type + " " + a.task + " " + a.group_id + " " + a.action_id} reason={reason} affect={a}/>
```

Comment:
JS has [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) which would make some of these concatenations a little easier on the eyes